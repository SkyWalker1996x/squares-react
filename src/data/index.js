const initTableConfig = {
  width: 4,
  height: 4,
  cellSize: 50,
};

const initWidth = 4;
const initHeight = 4;
const initCellSize = 50;

const initTableInteractive = {
  active: false,
  cellIndex: 0,
  rowIndex: 0,
  offsetLeft: 0,
  offsetTop: 0,
};

export {
  initTableConfig,
  initTableInteractive,
  initCellSize,
  initHeight,
  initWidth,
};
