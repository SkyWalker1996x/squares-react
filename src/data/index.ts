import { generateId } from '../utils';

const initTableButtons = [
  {
    id: generateId(),
    label: '+',
    type: 'add',
    element: 'row',
    className: 'button add-row',
  },
  {
    id: generateId(),
    label: '+',
    type: 'add',
    element: 'column',
    className: 'button add-col',
  },
  {
    id: generateId(),
    label: '-',
    type: 'remove',
    element: 'row',
    className: 'button remove-row',
  },
  {
    id: generateId(),
    label: '-',
    type: 'remove',
    element: 'column',
    className: 'button remove-col',
  },
];

const initWidth = 4;
const initHeight = 4;
const initCellSize = 50;

const initTableInteractive = {
  active: false,
  activeCellIndex: 0,
  activeRowIndex: 0,
  offsetLeft: 0,
  offsetTop: 0,
};

const delayTable = 500;

export {
  delayTable,
  initTableButtons,
  initTableInteractive,
  initCellSize,
  initHeight,
  initWidth,
};
