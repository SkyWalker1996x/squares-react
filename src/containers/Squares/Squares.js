import React, { useState } from 'react';
import { Table } from '../../components/Table/Table';
import { Button } from '../../components/Button/Button';
import { transformConfig, generateId } from '../../utils';
import { initTableInteractive, tableButtons } from '../../data';

const Squares = ({ width, height, cellSize }) => {
  const [tableConfig, setTableConfig] = useState(
    transformConfig({ width, height, cellSize })
  );
  const [tableInteractive, setTableInteractive] = useState(
    initTableInteractive
  );

  const addTableElement = (element) => {
    setTableConfig((prevState) => {
      const { rows, columns } = prevState;

      const newItem = {
        id: generateId(),
      };

      if (element === 'row') {
        return {
          ...prevState,
          rows: [...rows, newItem],
        };
      } else {
        return {
          ...prevState,
          columns: [...columns, newItem],
        };
      }
    });
  };
  const removeTableElement = (element) => {
    const { rows, columns } = tableConfig;
    if (element === 'row' && rows.length <= 1) return;
    if (element === 'column' && columns.length <= 1) return;

    if (element === 'row') {
      const { activeRowIndex: removeRowIndex } = tableInteractive;
      const newRows = rows.filter(
        (row, rowIndex) => rowIndex !== removeRowIndex
      );

      setTableConfig((prevState) => {
        return {
          ...prevState,
          rows: newRows,
        };
      });
    } else {
      const { activeCellIndex: removeCellIndex } = tableInteractive;
      const newColumns = columns.filter(
        (column, colIndex) => colIndex !== removeCellIndex
      );

      setTableConfig((prevState) => {
        return {
          ...prevState,
          columns: newColumns,
        };
      });
    }

    setTableInteractive((prevState) => {
      return {
        ...prevState,
        active: false,
      };
    });
  };
  const overTable = (e) => {
    if (e.target.className.includes('cell')) {
      setTableInteractive({
        active: true,
        activeCellIndex: e.target.cellIndex,
        activeRowIndex: e.target.parentNode.rowIndex,
        offsetLeft: e.target.offsetLeft,
        offsetTop: e.target.offsetTop,
      });
    } else {
      setTableInteractive((prevState) => {
        return {
          ...prevState,
          active: true,
        };
      });
    }
  };
  const outTable = () => {
    setTableInteractive((prevState) => {
      return {
        ...prevState,
        active: false,
      };
    });
  };

  const styleRemoveColBtn = {
    display:
      tableConfig.columns.length < 2
        ? 'none'
        : tableInteractive.active
        ? 'block'
        : 'none',
    left: tableInteractive.offsetLeft,
  };
  const styleRemoveRowBtn = {
    display:
      tableConfig.rows.length < 2
        ? 'none'
        : tableInteractive.active
        ? 'block'
        : 'none',
    top: tableInteractive.offsetTop,
  };
  const styleCellSize = {
    width: tableConfig.cellSize,
    height: tableConfig.cellSize,
  };

  const buttons = tableButtons.map((btnConfig) => {
    return (
      <Button
        addTableElement={addTableElement}
        removeTableElement={removeTableElement}
        btnConfig={btnConfig}
        styleRemoveRowBtn={styleRemoveRowBtn}
        styleRemoveColBtn={styleRemoveColBtn}
      />
    );
  });

  return (
    <div
      className="table-wrapper"
      onMouseOver={overTable}
      onMouseLeave={outTable}
    >
      <Table tableConfig={tableConfig} styleCellSize={styleCellSize} />
      {buttons}
    </div>
  );
};

export default Squares;
