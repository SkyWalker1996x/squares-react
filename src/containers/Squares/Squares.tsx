import React, { useState } from 'react';
import Table from '../../components/Table/Table';
import Button from '../../components/Button/Button';
import { transformConfig, generateId } from '../../utils';
import { initTableInteractive, initTableButtons, delayTable } from '../../data';
import {
  SquaresProps,
  TableInteractive,
  TableConfig,
  TimeoutTable,
  BtnConfig,
  StyleRemoveBtnType,
  StyleCellSizeType,
  TableElementsListeners
} from '../../interfaces';

const Squares = ({ width, height, cellSize }: SquaresProps) => {
  let timeoutTable: TimeoutTable;

  const [tableConfig, setTableConfig] = useState<TableConfig>(
    transformConfig({ width, height, cellSize })
  );
  const [tableInteractive, setTableInteractive] = useState<TableInteractive>(
    initTableInteractive
  );

  const addTableElement: TableElementsListeners = (element) => {
    setTableConfig((prevState) => {
      const { rows, columns } = prevState;

      const newItem = {
        id: generateId(),
      };

      if (element === 'row') {
        return {
          ...prevState,
          rows: [...rows, newItem],
        };
      } else {
        return {
          ...prevState,
          columns: [...columns, newItem],
        };
      }
    });
  };
  const removeTableElement: TableElementsListeners = (element) => {
    const { rows, columns } = tableConfig;
    if (element === 'row' && rows.length <= 1) return;
    if (element === 'column' && columns.length <= 1) return;

    if (element === 'row') {
      const { activeRowIndex: removeRowIndex } = tableInteractive;
      const newRows = rows.filter(
        (row, rowIndex) => rowIndex !== removeRowIndex
      );

      setTableConfig((prevState) => {
        return {
          ...prevState,
          rows: newRows,
        };
      });
    } else {
      const { activeCellIndex: removeCellIndex } = tableInteractive;
      const newColumns = columns.filter(
        (column, colIndex) => colIndex !== removeCellIndex
      );

      setTableConfig((prevState) => {
        return {
          ...prevState,
          columns: newColumns,
        };
      });
    }

    setTableInteractive((prevState) => {
      return {
        ...prevState,
        active: false,
      };
    });
  };
  const overTable = (e: React.MouseEvent) => {
    clearTimeout(timeoutTable);

    if ((e.target as HTMLElement).className.includes('cell')) {
      setTableInteractive({
        active: true,
        activeCellIndex: (e.target as HTMLTableCellElement).cellIndex,
        offsetLeft: (e.target as HTMLTableElement).offsetLeft,
        offsetTop: (e.target as HTMLTableElement).offsetTop,
        // @ts-ignore
        activeRowIndex: (e.target as HTMLTableElement).parentNode.rowIndex,
      });
    } else {
      setTableInteractive((prevState) => {
        return {
          ...prevState,
          active: true,
        };
      });
    }
  };
  const outTable = () => {
    timeoutTable = setTimeout(() => {
      setTableInteractive((prevState) => {
        return {
          ...prevState,
          active: false,
        };
      });
    }, delayTable);
  };

  const styleRemoveColBtn: StyleRemoveBtnType = {
    display:
      tableConfig.columns.length < 2
        ? 'none'
        : tableInteractive.active
        ? 'block'
        : 'none',
    left: tableInteractive.offsetLeft,
  };
  const styleRemoveRowBtn: StyleRemoveBtnType = {
    display:
      tableConfig.rows.length < 2
        ? 'none'
        : tableInteractive.active
        ? 'block'
        : 'none',
    top: tableInteractive.offsetTop,
  };
  const styleCellSize: StyleCellSizeType = {
    width: tableConfig.cellSize,
    height: tableConfig.cellSize,
  };

  const buttons = initTableButtons.map((btnConfig: BtnConfig) => {
    const {id, label, element, className, type} = btnConfig;
    console.log(btnConfig);

    return (
      <Button
        key={id}
        onClick={
          type === 'remove'
              ? () => removeTableElement(element)
              : () => addTableElement(element)
        }
        label={label}
        className={className}
        style={
          type === 'add'
              ? {}
              : element === 'row'
              ? styleRemoveRowBtn
              : styleRemoveColBtn
        }
      />
    );
  });

  return (
    <div
      className="table-wrapper"
      onMouseOver={overTable}
      onMouseLeave={outTable}
    >
      <Table tableConfig={tableConfig} styleCellSize={styleCellSize} />
      {buttons}
    </div>
  );
};

export default Squares;
