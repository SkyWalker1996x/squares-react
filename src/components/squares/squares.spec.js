import React from 'react';
import { render } from 'enzyme';
import Squares from './index';
import { initWidth, initHeight, initCellSize } from '../../data';

describe('Squares testing', () => {
  describe('Render a table', () => {
    const testProps = {
      width: initWidth,
      height: initHeight,
      cellSize: initCellSize,
    };
    const testSquares = render(<Squares {...testProps} />);

    it('table render with initial props', () => {
      expect(testSquares).toMatchSnapshot();
    });

    it('number of rows into a table should be 4', () => {
      const table = testSquares.children()[0].children[0];
      expect(table.children).toHaveLength(initHeight);
    });

    it('number of cells into a row should be 4', () => {
      const table = testSquares.children()[0].children[0];
      const rows = table.children;
      rows.forEach((node) => {
        expect(node.children).toHaveLength(initWidth);
      });
    });
  });
});
