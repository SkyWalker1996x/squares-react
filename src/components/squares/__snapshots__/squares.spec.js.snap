// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Squares testing Render a table table render with initial props 1`] = `
<div
  class="table-wrapper"
>
  <table
    class="table"
  >
    <tbody>
      <tr
        class="row"
      >
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
      </tr>
      <tr
        class="row"
      >
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
      </tr>
      <tr
        class="row"
      >
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
      </tr>
      <tr
        class="row"
      >
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
        <td
          class="cell"
          style="width:50px;height:50px"
        />
      </tr>
    </tbody>
  </table>
  <button
    class="button add-row"
  >
    +
  </button>
  <button
    class="button add-col"
  >
    +
  </button>
  <button
    class="button remove-row"
    style="display:none;top:0"
  >
    -
  </button>
  <button
    class="button remove-col"
    style="display:none;left:0"
  >
    -
  </button>
</div>
`;
